Add LIP "Introduce BFT module"

janhack opened the pull request.

MaximeGagnebin reviewed (approved):
I only left a few typo style comments and updates for the merged LIPs. Nice job!

janhack commented on proposals/lip-introduce_bft_module.md:
````diff
@@ -0,0 +1,775 @@
+##### Execution
+
+```python
+areDistinctHeadersContradicting(blockHeaderInfo1,blockHeaderInfo2):
````
Fixed.

janhack commented on proposals/lip-introduce_bft_module.md:
````diff
@@ -0,0 +1,775 @@
+##### Execution
+
+```python
+areHeadersContradicting(blockHeader1,blockHeader2):
````
Fixed.

janhack commented on proposals/lip-introduce_bft_module.md:
````diff
@@ -0,0 +1,775 @@
+
+```python
+getBFTParametersInternal(h):
+    for key,value in BFT Parameters substore ordered
````
Fixed.

janhack replied:
Fixed.

janhack commented on proposals/lip-introduce_bft_module.md:
```diff
@@ -0,0 +1,775 @@
+    if  blockHeader1.blockID == blockHeader2.blockID:
+        # no contradiction, as block headers are the same
+        return False
+    return areDistinctHeadersContradicting(blockHeader1,blockHeader2)
```
Fixed.

janhack commented on proposals/lip-introduce_bft_module.md:
```diff
@@ -0,0 +1,775 @@
+[lip-certificate-generation]: https://research.lisk.com/t/introduce-certificate-generation-mechanism/296
+[lip-dpos]: https://research.lisk.com/t/define-state-and-state-transitions-of-dpos-module/320
+[lip-poa]: https://research.lisk.com/t/proof-of-authority-validator-selection-mechanism/288
+[lip-reward-module]: https://research.lisk.com/t/define-state-transitions-of-reward-module/316
```
Fixed.

janhack commented on proposals/lip-introduce_bft_module.md:
```diff
@@ -0,0 +1,775 @@
+
+[lip-certificate-generation]: https://research.lisk.com/t/introduce-certificate-generation-mechanism/296
+[lip-dpos]: https://research.lisk.com/t/define-state-and-state-transitions-of-dpos-module/320
+[lip-poa]: https://research.lisk.com/t/proof-of-authority-validator-selection-mechanism/288
```
Fixed.

janhack commented:
I addressed the review comments and there were some changes resulting from the BFT module implementation, see https://github.com/LiskHQ/lisk-sdk/issues/6742. I therefore re-requested the reviews.

ricott1 commented on proposals/lip-introduce_bft_module.md:
```diff
@@ -416,55 +416,55 @@ computeValidatorsHashInternal(validators, certificateThreshold):
+#### getBFTParametersInternal
 
-This function creates a `blockBFTInfo` object with a subset of the block properties that are required for the BFT module.
+This function is a convenience function for obtaining the BFT parameters for a given height.
```
```suggestion
This function is a convenience function for obtaining the BFT parameters at a given height.
```

Minor, later you use 'at the height', I think is more correct.

MaximeGagnebin reviewed (approved):
LGTM

AndreasKendziorra commented on proposals/lip-introduce_bft_module.md:
```diff
@@ -0,0 +1,805 @@
+
+1. The object `getBlockBFTProperties(b.header)` is added to the beginning of the array `bftVotes.blockBFTInfos`.
+2. If `bftVotes.blockBFTInfos[MAX_LENGTH_BLOCK_BFT_INFOS]` exists, then the corresponding value is removed from the array.
+3. The function `applyPrevotesPrecommits` from the LIP ["Add weights to Lisk-BFT consensus protocol"][lip-weighted-lisk-bft] is executed to update the prevote and precommit weights.
```
I just spotted by chance that the function name in the other LIP is `updatePrevotesPrecommits` instead of `applyPrevotesPrecommits`.

janhack replied:
Fixed, thanks!

shuse2 commented on proposals/lip-introduce_bft_module.md:
```diff
@@ -0,0 +1,805 @@
+    if length(bftVotes.blockBFTInfos) > 0:
+        nextHeight = bftVotes.blockBFTInfos[0].height + 1
+    else:
+        # only directly after processing the genesis block we have
+        # length(bftVotes.blockBFTInfos) == 0
+        nextHeight = bftVotes.maxHeightPrevoted + 1
```
when this is called during genesis block creation or execution, (for example h = 0), it will set this `nextHeight` to `1`. 
to get the `validatorsHash` for the genesis block, it will call `getBFTParameters` with `h = 0` and it will fail because there is no parameter for `h = 0`

shuse2 replied:
The problem was when calling `getBFTParameters`, it needed to be called with h + 1 as discussed. There is no issue on the specification.
